drivers/pwm/pwm_gecko.c

```diff
@@ -0,0 +1,132 @@
+/*
+ * Copyright (c) 2021 Sun Amar.
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+#define DT_DRV_COMPAT silabs_gecko_pwm
+
+#include <drivers/pwm.h>
+#include <dt-bindings/pwm/pwm.h>
+#include "em_cmu.h"
+#include "em_timer.h"
+
+/** PWM configuration. */
+struct pwm_gecko_config {
+	TIMER_TypeDef *timer;
+	CMU_Clock_TypeDef clock;
+	TIMER_Prescale_TypeDef prescale_enum;
+	uint8_t channel;
+	uint8_t location;
+	uint8_t port;
+	uint8_t pin;
+};
+
+struct pwm_gecko_data {
+	int prescaler;
+};
+
+#define DEV_CFG(dev) \
+	((const struct pwm_gecko_config * const)(dev)->config)
+#define DEV_DATA(dev) \
+	((struct pwm_gecko_data *const)(dev)->data)
+
+static int pwm_gecko_pin_set(const struct device *dev, uint32_t pwm,
+			     uint32_t period_cycles, uint32_t pulse_cycles,
+			     pwm_flags_t flags)
+{
+	TIMER_InitCC_TypeDef compare_config = TIMER_INITCC_DEFAULT;
+	const struct pwm_gecko_config *cfg = DEV_CFG(dev);
+
+	if (BUS_RegMaskedRead(&cfg->timer->CC[pwm].CTRL,
+		_TIMER_CC_CTRL_MODE_MASK) != timerCCModePWM) {
+
+#ifdef _TIMER_ROUTE_MASK
+		BUS_RegMaskedWrite(&cfg->timer->ROUTE,
+			_TIMER_ROUTE_LOCATION_MASK,
+			cfg->location << _TIMER_ROUTE_LOCATION_SHIFT);
+		BUS_RegMaskedSet(&pwm->timer->ROUTE, 1 << pwm);
+#elif defined(_TIMER_ROUTELOC0_MASK)
+		BUS_RegMaskedWrite(&cfg->timer->ROUTELOC0,
+			_TIMER_ROUTELOC0_CC0LOC_MASK << (pwm * _TIMER_ROUTELOC0_CC1LOC_SHIFT),
+			cfg->location << (pwm * _TIMER_ROUTELOC0_CC1LOC_SHIFT));
+		BUS_RegMaskedSet(&cfg->timer->ROUTEPEN, 1 << pwm);
+#else
+#error Unsupported device
+#endif
+
+		compare_config.mode = timerCCModePWM;
+		TIMER_InitCC(cfg->timer, pwm, &compare_config);
+	}
+
+	cfg->timer->CC[pwm].CTRL |= (flags & PWM_POLARITY_INVERTED) ?
+		TIMER_CC_CTRL_OUTINV : 0;
+
+	TIMER_TopSet(cfg->timer, period_cycles);
+
+	TIMER_CompareBufSet(cfg->timer, pwm, pulse_cycles);
+
+	return 0;
+}
+
+static int pwm_gecko_get_cycles_per_sec(const struct device *dev,
+					uint32_t pwm,
+					uint64_t *cycles)
+{
+	const struct pwm_gecko_config *cfg = DEV_CFG(dev);
+	struct pwm_gecko_data *const data = DEV_DATA(dev);
+
+	*cycles = CMU_ClockFreqGet(cfg->clock) / data->prescaler;
+
+	return 0;
+}
+
+static const struct pwm_driver_api pwm_gecko_driver_api = {
+	.pin_set = pwm_gecko_pin_set,
+	.get_cycles_per_sec = pwm_gecko_get_cycles_per_sec
```

Comment:
Any chance of seeing support for PWM capture as well?

Reply:
I don`t want to postpone this PR because it may take me some time to be able to work on this part.
I`ll open another PR if I`ll do it in the future.